cranelift/codegen/src/machinst/buffer.rs

```diff
@@ -1077,15 +1080,34 @@ impl<I: VCodeInst> MachBuffer<I> {
             align,
             label
         );
-        let deadline = self.cur_offset().saturating_add(max_distance);
-        self.island_worst_case_size += data.len() as CodeOffset;
-        self.island_worst_case_size =
-            (self.island_worst_case_size + I::LabelUse::ALIGN - 1) & !(I::LabelUse::ALIGN - 1);
+        self.update_deadline(data.len(), max_distance);
         self.pending_constants.push(MachLabelConstant {
             label,
             align,
             data: SmallVec::from(data),
         });
+    }
+
+    /// Emit a trap instruction at some point in the future, binding the given
+    /// label to its offset. The trap will be placed at most `max_distance`
+    /// from the current offset.
+    pub fn defer_trap(&mut self, code: TrapCode, stack_map: Option<StackMap>) -> MachLabel {
```

Comment:
`defer_trap` doesn't have the `max_distance` argument that its doc-comment refers to. Do we need a max-distance for branch instructions on any target? Also it isn't "the given label", since it returns the label. I assume the comment was just copied from `defer_constant`.

As future work, could `defer_trap` deduplicate the trap pool by returning the same `MachLabel` when given the same `TrapCode`? (At least if `stack_map` is `None`.) Or do we need to know exactly which check trapped?

Reply:
Aha you've reminded me of something I forgot in this PR which was to track the source location of traps to get the correct file origin information when a trap happens. Previously this would naturally happen due to the source location tracking but by moving these to the end the tracking is lost. It turns out we don't have anything in the wasm test suite testing this just yet! I've now added in source location tracking for deferred traps in addition to some more tests which require the logic to be here.

So now to get back to your question, about deduplicating, maybe! The source location adds another vector by which this needs to be considered for deduplication, and it's pretty unlikely to get deduplicated with the source information (except for perhaps some of those x64 lowerings of float-to-int conversions which have a bunch of traps). With source locations considered too, though, I think it's safe to deduplicate in the future.